packages/compass-connections-navigation/src/virtual-list/virtual-list.tsx

```diff
@@ -0,0 +1,217 @@
+  renderItem: RenderItem<T>;
+  getItemKey?: (item: T) => string;
+  onDefaultAction: OnDefaultAction<NotPlaceholderTreeItem<T>>;
+  onExpandedChange: OnExpandedChange<NotPlaceholderTreeItem<T>>;
+
+  __TEST_OVER_SCAN_COUNT?: number;
+};
+
+const navigationTree = css({
+  flex: '1 0 auto',
+});
+
+function useAction<T>(fn: RenderItem<T>): RenderItem<T> {
+  const ref = useRef(fn);
+  ref.current = fn;
+  return useMemo(() => {
+    return (props) => ref.current(props);
+  }, []);
+}
+
+export function VirtualTree<T extends VirtualItem>({
+  dataTestId,
+  activeItemId,
+  items,
+  width,
+  height,
+  itemHeight,
+  getItemKey: _getItemKey,
+  renderItem: _renderItem,
+  onDefaultAction,
+  onExpandedChange,
+  __TEST_OVER_SCAN_COUNT,
+}: VirtualTreeProps<T>) {
+  const listRef = useRef<List | null>(null);
+  const renderItem = useAction(_renderItem);
+  const onFocusMove = useCallback(
+    (item: VirtualTreeItem) => {
+      const idx = items.findIndex(
+        (i) => !isPlaceholderItem(i) && i.id === item.id
+      );
+      if (idx >= 0) {
+        // It is possible that the item we are trying to move the focus to is
+        // not rendered currently. Scroll it into view so that it's rendered and
+        // can be focused
+        listRef.current?.scrollToItem(idx);
+      }
+    },
+    [items]
+  );
+  const [rootProps, currentTabbable] = useVirtualNavigationTree<HTMLDivElement>(
+    {
+      items,
+      activeItemId,
+      onExpandedChange,
+      onFocusMove,
+    }
+  );
+
+  const id = useId();
+
+  const itemData = useMemo(() => {
+    return {
+      items,
+      currentTabbable,
+      renderItem,
+      onDefaultAction,
+    };
+  }, [items, renderItem, currentTabbable, onDefaultAction]);
+
+  const getItemKey = useCallback(
+    (index: number, data: VirtualItemData<T>) => {
+      if (!_getItemKey) {
+        return index;
+      }
+      return _getItemKey(data.items[index]);
+    },
+    [_getItemKey]
+  );
+
+  return (
+    <div
+      data-testid={dataTestId}
+      role="tree"
+      aria-labelledby={id}
+      className={navigationTree}
+      {...rootProps}
+    >
+      <List<VirtualItemData<T>>
+        ref={listRef}
+        width={width}
+        height={height}
+        itemData={itemData}
+        itemCount={items.length}
+        itemSize={itemHeight}
+        itemKey={getItemKey}
+        overscanCount={__TEST_OVER_SCAN_COUNT ?? 5}
+      >
+        {TreeItem}
+      </List>
+    </div>
+  );
+}
+
+type VirtualItemData<T extends VirtualItem> = {
+  items: T[];
+  currentTabbable?: string;
+  renderItem: RenderItem<T>;
+  onDefaultAction: OnDefaultAction<NotPlaceholderTreeItem<T>>;
+};
+function TreeItem<T extends VirtualItem>({
+  index,
+  data,
+  style,
+}: ListChildComponentProps<VirtualItemData<T>>) {
+  const { renderItem, items } = data;
+  const item = useMemo(() => items[index], [items, index]);
+  const focusRingProps = useFocusRing();
+
+  const Component = useMemo(() => {
```

Comment:
Nit: uppercase is usually used for component constructors, this is already an element that you "rendered"
```suggestion
  const component = useMemo(() => {
```